test/test_integration_cluster.rb

```diff
@@ -112,117 +91,166 @@ def test_term_worker_clean_exit
   def test_stuck_external_term_spawn
-    worker_respawn { |l, phase0_worker_pids|
-      phase0_worker_pids.each { |p| Process.kill :TERM, p }
-    }
+    skip_unless_signal_exist? :TERM
+
+    worker_respawn(0) do |phase0_worker_pids|
+      last = phase0_worker_pids.last
+      phase0_worker_pids.each do |pid|
+        Process.kill :TERM, pid
+        sleep 4 unless pid == last
+      end
+    end
   end
 
   # mimicking stuck workers, test restart
   def test_stuck_phased_restart
-    worker_respawn { |l, phase0_worker_pids| l.phased_restart }
+    skip_unless_signal_exist? :USR1
+    worker_respawn { |phase0_worker_pids| Process.kill :USR1, @pid }
   end
 
   private
 
-  def worker_respawn
-    skip NO_FORK_MSG unless HAS_FORK
-    port = UniquePort.call
-    workers_booted = 0
-
-    conf = Puma::Configuration.new do |c|
-      c.bind "tcp://#{HOST}:#{port}"
-      c.threads 1, 1
-      c.workers WORKERS
-      c.worker_shutdown_timeout 2
-      c.app TestApps::SLEEP
-      c.after_worker_fork { |idx| workers_booted += 1 }
-    end
+  # Send requests 10 per second.  Send 10, then :TERM server, then send another 30.
+  # No more than 10 should throw Errno::ECONNRESET.
+  def term_closes_listeners(unix: false)
+    skip_unless_signal_exist? :TERM
+
+    cli_server "-w #{WORKERS} -t 0:6 -q test/rackup/sleep_pid.ru", unix: unix
+    threads = []
+    replies = []
+    mutex = Mutex.new
+    div   = 10
 
-    # start Puma via launcher
-    thr, launcher, _e = run_launcher conf
+    refused = thread_run_refused unix: unix
 
-    # make sure two workers have booted
-    time = 0
-    until workers_booted >= WORKERS || time >= 10
-      sleep 2
-      time += 2
+    41.times.each do |i|
+      if i == 10
+        threads << Thread.new do
+          sleep i.to_f/div
+          Process.kill :TERM, @pid
+          mutex.synchronize { replies << :term_sent }
+        end
+      else
+        threads << Thread.new do
+          thread_run replies, i.to_f/div, 1, mutex, refused, unix: unix
+        end
+      end
     end
 
-    cluster = launcher.instance_variable_get :@runner
+    threads.each(&:join)
 
-    http0 = Net::HTTP.new HOST, port
-    http1 = Net::HTTP.new HOST, port
-    body0 = nil
-    body1 = nil
+    responses = replies.count { |r| r[/\ASlept 1/] }
+    resets    = replies.count { |r| r == :reset    }
+    refused   = replies.count { |r| r == :refused  }
+    msg = "#{responses} responses, #{resets} resets, #{refused} refused"
 
-    worker0 = Thread.new do
-      begin
-        req0 = Net::HTTP::Get.new "/sleep35", {}
-        http0.start.request(req0) { |rep0| body0 = rep0.body }
-      rescue
-      end
-    end
+    assert_operator 9,  :<=, responses, msg
 
-    worker1 = Thread.new do
-      begin
-        req1 = Net::HTTP::Get.new "/sleep40", {}
-        http1.start.request(req1) { |rep1| body1 = rep1.body }
-      rescue
+    assert_operator 10, :>=, resets   , msg
+
+    assert_operator 20, :<=, refused  , msg
```

Comment:
This feels like a pretty big change. Wasn't the original test that _all_ responses must be refused?

Reply:
I may be incorrect, but I think the updated test is probably more accurate.  GIven that the system may be loaded down, resource starved, parallel tests, etc, it will take some finite amount of time to shut down.

I figured this would define a baseline for that time, it may be able to be improved, but it certainly will show if the code if changed and the time is affected for the worse.

Reply:
I still can't quite get past this. This test wasn't failing intermittently on Linux/Windows before, so I don't like that it is (?) now and we need this "<" stuff.